Simplify Vision class

PeterJCLaw opened the pull request:
Separate wrapping of the Apriltag library out from the `Vision` class so that the library lifetimes are all in one place and so the actual intent of the processing is simpler to read.

This also greatly simplifies the initialisation logic of the `Vision` class by moving to lazy properties.

PeterJCLaw committed:
Simplify apriltag wrapping logic by isolating it

Move native wrapping code into its own class which just wraps the
lib and ffi portions with a more pythonic interface. So far this
keeps the previous initialisation and logic flows around, but the
result substantially simplifies the image processing logic within
the `Vision` class.

PeterJCLaw committed:
Use direct lazy initialisation for the camera and apriltag library

This moves the initialisation of these two behind properties rather
than an explicit method call. This _greatly_ simplifies the code
and means that the ordering of the initialisations no longer needs
to be explicitly encoded -- it will just always be correct.

PeterJCLaw committed:
Fix some lint issues

prophile commented on sb_vision/native/apriltag/__init__.py:
```diff
@@ -1 +1,98 @@
+            lib.apriltag_detector_destroy(self._detector)
+
+        if self._working_image:
+            lib.image_u8_destroy(self._working_image)
```
I'm not wild about `__del__` for resource management, because Python makes no guarantees about when (or even whether) `__del__` is called.

Would be up for a context manager/explicit `close` to do this.

PeterJCLaw replied:
c97a296

prophile commented on sb_vision/native/apriltag/__init__.py:
```diff
@@ -1 +1,98 @@
+                self.image_size,
+                img.size,
+            )
+        )
```
Would be more in favour of a `ValueError` in this case rather than using `assert`.

`assert`s can be compiled out when Python is run with `-O`, and this isn't an _internal_ assertion about consistent state: from a user's point of view passing an image of the incorrect size is something that _can_ happen.

PeterJCLaw replied:
0571a47

prophile commented:
Would you kindly fix the linting errors and we can merge this :)

PeterJCLaw committed:
This is a detector

PeterJCLaw committed:
Convert assertion to ValueError check

PeterJCLaw committed:
Isort

PeterJCLaw committed:
Wrap for line length

PeterJCLaw committed:
Fix docstrings

PeterJCLaw committed:
Lint

PeterJCLaw committed:
Unwrap docstring

PeterJCLaw committed:
Add explicit mechanisms for lifetime control

This introduces a close method as well as context manager support
for the AprilTagDetector class so that users have more direct
control over when it disposes of its native resources.

PeterJCLaw commented:
Hnng, apparently I can't leave the TODO note in place, yet I'd very much intended to do so -- it's a longer term thing rather than something for this PR. Would it be better to convert it to an issue?

PeterJCLaw committed:
Convert TODO to note

PeterJCLaw commented:
Raised TODO as #17.

Adimote commented:
In a perfect world, we'd allow TODOs but only if they were of the form `//TODO(#17)` where `#17` is the ID of an issue. Have created a trello card for this: https://trello.com/c/TJUQIGNO

Adimote committed:
Merge branch 'master' into vision-raii

PeterJCLaw committed:
Tidy imports

PeterJCLaw commented:
@prophile linting issues fixed; could you take another look?

PeterJCLaw commented:
I can't remember whether I've full-stack tested this; as it's been a while since I created this PR I'll do that before merging.
